lib/api/CInferenceModelDefinition.cc

```diff
@@ -380,6 +392,48 @@ std::string CInferenceModelDefinition::jsonString() {
     return resultString;
 }
 
+std::stringstream CInferenceModelDefinition::jsonStringCompressedFormat() const {
+    std::stringstream compressedStream;
+    using TFilteredOutput = boost::iostreams::filtering_stream<boost::iostreams::output>;
+    std::string modelDefinitionStr{jsonString()};
+    {
+        TFilteredOutput outFilter;
+        outFilter.push(boost::iostreams::gzip_compressor());
+        outFilter.push(core::CBase64Encoder());
+        outFilter.push(compressedStream);
+        outFilter << modelDefinitionStr;
+        outFilter.flush();
+    }
+    return compressedStream;
+}
+
+void CInferenceModelDefinition::addToDocumentCompressed(TRapidJsonWriter& writer) const {
+    std::stringstream compressedString{jsonStringCompressedFormat()};
```

Comment:
Ok this is probably fine, but again you are creating a copy of the entire model before chunking.

Chunking is a streaming operation: i.e. you need to buffer only one chunk then flush that chunk to the writer. It strikes me that this can be handled as a new stream sink: you just need to move the logic below (roughly) into a class with a `std::streamsize write(const char* s, std::streamsize n)` method. This doesn't actually feel like it would be a whole lot more complicated. It would be nice to make the actual logic to write a chunk pluggable. This could be done by having a callback to actually write the chunk.

We do have `CStateCompressor::CChunkFilter` I wonder if it can be factored out and generalised for this purpose. I would also be interested in @droberts195 thoughts on this.

Reply:
This was basically my biggest comment too.

There seems to be a fundamental mismatch in the way we are treating these model definitions in different places.  In some places we're saying they're so huge that they need compression and chunking, and in other places we're saying they're so small that you can just create a temporary copy without worrying about the memory usage.

One problem with the whole architecture is that the model definition is considered to be a type of result, not a type of state.  Results are going through the `CRapidJsonConcurrentLineWriter` which expects to take a series of small documents to be sent to the results named pipe in a thread safe way so results written from different threads don't corrupt each other.  State goes through a different named pipe and is designed to be streamed, such that entire state documents never exist in memory in their entirety.  But unfortunately the Java side is not set up to receive different types of documents from a single state pipe, so it would be hard to use it for both the state for restarts and the model definition.

So it seems like the most realistic goal would be to continue to use the `CRapidJsonConcurrentLineWriter` to write the compressed model as a single document, but use streaming internally within the C++ code to avoid ever having to create the complete uncompressed model in JSON format in memory.

> We do have `CStateCompressor::CChunkFilter`

I think something along similar lines should be used, but it's probably best to create a new bespoke class that does it for this type of document.  `CChunkFilter` is designed to work with the state persistence framework, which is already generic in that it can persist to both XML and JSON, and was designed to be able to stream that state to a variety of stores, which have varied from HTTP to named pipes over the years.  To try and make it even more generic is going to make it unbelievably complex.  So I would suggest using some of the same ideas but creating a different class to do it.

Reply:
I think in that case, my inclination would be to make the new sink very bespoke and just lift the actual writing out of the while loop below into the sink write method and have it local to this cc file.